src/main/java/org/folio/edge/sip2/MainVerticle.java

```diff
@@ -103,148 +108,190 @@ public void start(Promise<Void> startFuture) {
+              tenantConfig.getString("charset", "IBM850"));
+          final String messageDelimiter = tenantConfig.getString("messageDelimiter", "\r");
+
+          socket.handler(RecordParser.newDelimited(messageDelimiter, buffer -> {
+            final Timer.Sample sample = metrics.sample();
+
+            if (Objects.isNull(sessionData.getTenant())) {
+              log.error("No tenant configured for address: {}  message ignored.", clientAddress);
+              return;
+            }
+
+            final String messageString = buffer.getString(0, buffer.length(),
+                sessionData.getCharset());
+
+            Command command = UNKNOWN;
+
+            try {
+              final Parser parser = Parser.builder()
+                  .delimiter(sessionData.getFieldDelimiter())
+                  .charset(Charset.forName(sessionData.getCharset()))
+                  .errorDetectionEnabled(sessionData.isErrorDetectionEnabled())
+                  .timezone(sessionData.getTimeZone())
+                  .build();
+
+              //parsing
+              final Message<Object> message = parser.parseMessage(messageString);
+
+              command = message.getCommand();
+
+              //process validation results
+              if (!message.isValid()) {
+                log.error("Message is invalid: {}", messageString);
+                handleInvalidMessage(message, socket, sessionData, messageDelimiter, sample,
+                    metrics);
+                return;
+              }
+
+              //check if the previous message needs resending
+              if (requiredResending(sessionData, message)) {
+                resendPreviousMessage(sessionData, sample,
+                    metrics, socket, command);
+                return;
+              }
+
+              ISip2RequestHandler handler = handlers.get(command);
+
+              if (handler == null) {
+                log.error("Error locating handler for command {}", command.name());
+                sample.stop(metrics.commandTimer(command));
+                return;
+              }
+
+              executeHandler(message,
+                  sessionData, messageDelimiter,
+                  handler, sample,
+                  socket, metrics);
+            } catch (Exception ex) {
+              String message = "Problems handling the request: " + ex.getMessage();
+              log.error(message, ex);
+              // Return an error message for now for the sake of negative testing.
+              // Will find a better way to handle negative test cases.
+              sample.stop(metrics.commandTimer(command));
+              socket.write(message + messageDelimiter, sessionData.getCharset());
+              metrics.requestError();
+            }
+          }));
+          socket.exceptionHandler(t -> {
+            log.info("Socket exceptionHandler caught an issue, see error logs for more details");
+            log.error("Socket exception", t);
```

Comment:
Think the two log lines are redundant. Probably just do `log.error` here.

Reply:
done